assets/src/edit-story/utils/context.js

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { shallowEqual } from 'react-pure-render';
+import { useCallback, useRef } from 'react';
+import { useContextSelector } from 'use-context-selector';
+
+export { createContext, useContext } from 'use-context-selector';
+
+/**
+ * This hook returns a part of the context value by selector.
+ * It will trigger a re-render only if the selected value has changed.
+ *
+ * @param {React.Context} context
+ * @param {Function} selector Returns the part of the context that the
+ * consumer is interested in.
+ * @param {Function} equalityFn Used to compare the selected context value.
+ * If the context fragment has not changed, a re-render will not be triggered.
+ * @return {*}
+ */
+export const useContextSelectorWithEqFn = (
+  context,
+  selector,
+  { equalityFn }
+) => {
+  const ref = useRef();
+  const patchedSelector = equalityFn
+    ? useCallback(
+        (state) => {
+          const selected = selector(state);
+          if (equalityFn(ref.current, selected)) {
+            return ref.current;
+          }
+          ref.current = selected;
+          return selected;
+        },
+        [selector, equalityFn]
+      )
+    : selector;
+  return useContextSelector(context, patchedSelector);
+};
+
+/**
+ * This hook returns a part of the context value by selector.
```

Comment:
A bit of a nit, but it seems to be that all we want is useContextSelector(context, selector, equalityFn=shallowEqual)? Is it a real use case for us to use useContextSelectorWithEqFn w/o any kind of equality function? Would it just be like a plain useContext() then?

Reply:
I chose to add 'useContextSelectorShallow' for a few reasons:
a) In case useContextSelector from use-context-selector is imported by mistake 🙂 this would introduce hard to find defects.
b) To make the shallow eq behavior clear at the callsite, otherwise it's very opaque that individual object attributes are memoized.
c) useContextSelector has a well-defined behavior from use-context-selector, and shadowing that with the same name and shallow-eq behavior may be unclear.

Point a) could be mitigated by changing the name, eg. 'useContextSelectorEq'.

Reply:
> In case `useContextSelector` from `use-context-selector` is imported by mistake

We could easily prevent that on a linter level too, FWIW

Reply:
I agree. Linter is the right way to go for the package import confusions. Probably needs to be done no matter what :) Please consider that and make a decision on naming. I'm otherwise good leaving this in your hands. And it's easy enough to rename if we don't like at some point.

Reply:
I'll change it to be 'useContextSelector'.

It would be useful to add a linter for package import confusion, but a developer could still reads the docs for use-context-selector and make a bad assumption.